miio/device.py

```diff
@@ -252,31 +284,21 @@ def actions(self) -> Dict[str, ActionDescriptor]:
 
     def settings(self) -> Dict[str, SettingDescriptor]:
         """Return device settings."""
-        settings = self.status().settings()
-        for setting in settings.values():
-            # TODO: Bind setter methods, this should probably done only once during init.
-            if setting.setter is None:
-                # TODO: this is ugly, how to fix the issue where setter_name is optional and thus not acceptable for getattr?
-                if setting.setter_name is None:
-                    raise Exception(
-                        f"Neither setter or setter_name was defined for {setting}"
-                    )
-
-                setting.setter = getattr(self, setting.setter_name)
-            if (
-                isinstance(setting, EnumSettingDescriptor)
-                and setting.choices_attribute is not None
-            ):
-                retrieve_choices_function = getattr(self, setting.choices_attribute)
-                setting.choices = retrieve_choices_function()  # This can do IO
+        if self._settings is None:
+            raise Exception(
+                "setting descriptors schould first be cached using the info() call"
+            )
 
-        return settings
+        return self._settings
 
     def sensors(self) -> Dict[str, SensorDescriptor]:
         """Return device sensors."""
-        # TODO: the latest status should be cached and re-used by all meta information getters
-        sensors = self.status().sensors()
-        return sensors
+        if self._sensors is None:
+            raise Exception(
+                "sensor descriptors schould first be cached using the info() call"
```

Comment:
I think it would be better just to call `_initialize_descriptors` considering it's just a single method call that will return fast if already initialized.

Reply:
Well I was doing it this way to make sure accesing the descriptors will not do IO.
For HomeAssistant this is important because we can not do IO inside the event loop.
So we need to schedule IO calls in the executer.

In this way we do not need to worry about this going wrong.

But if we are carefull in HomeAssistant about always calling .info() before using any of the descriptors we could indeed do what you suggest and just call the initialization, but I suspect this might cause a bug in the future if someone forgets.

Reply:
What do you think?

Reply:
So, I think it's a common design pattern (at least among homeassistant folks?) to avoid doing I/O on property accesses to make it clear it's fine to access these anywhere. As these are currently just regular methods, I think we should try to keep the API simple for now. We can discuss about the pros and cons on converting these to properties another time, but let's keep it as simple as possible for now.

Reply:
done